lucene/core/src/java/org/apache/lucene/search/Matches.java

```diff
@@ -0,0 +1,161 @@
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.util.BytesRef;
+
+/**
+ * Reports the positions and optionally offsets of all matching terms in a query
+ * for a single document
+ *
+ * To find all fields that have matches, call {@link #getMatchFields()}
+ *
+ * To obtain a {@link MatchesIterator} for a particular field, call {@link #getMatches(String)}
+ */
+public class Matches {
+
+  private final Map<String, MatchesIterator> matches;
+
+  /**
+   * Create a simple {@link Matches} for a single field
+   */
+  public static Matches fromField(String field, MatchesIterator it) {
+    if (it == null) {
+      return null;
+    }
+    return new Matches(field, it);
+  }
+
+  /**
+   * Create an empty {@link Matches} for a Weight
+   *
+   * If the Weight's parent query does not match this document, returns {@code null},
+   * otherwise returns a {@link Matches} document with an empty iterator on the given
+   * fields
+   */
+  public static Matches emptyMatches(LeafReaderContext context, int doc, Weight weight, String... fields) throws IOException {
+    Scorer scorer = weight.scorer(context);
+    if (scorer == null || scorer.iterator().advance(doc) != doc) {
+      return null;
+    }
+    List<Matches> matches = new ArrayList<>();
+    for (String field : fields) {
+      matches.add(Matches.fromField(field, EMPTY));
+    }
+    return Matches.fromSubMatches(matches);
+  }
+
+  /**
+   * Amalgamate a collection of {@link Matches} into a single object
+   */
+  public static Matches fromSubMatches(List<Matches> subMatches) throws IOException {
+    if (subMatches == null || subMatches.size() == 0) {
+      return null;
+    }
+    if (subMatches.size() == 1) {
+      return subMatches.get(0);
+    }
+    Map<String, MatchesIterator> matches = new HashMap<>();
+    Set<String> allFields = new HashSet<>();
+    for (Matches m : subMatches) {
+      allFields.addAll(m.getMatchFields());
+    }
+    for (String field : allFields) {
+      List<MatchesIterator> mis = new ArrayList<>();
+      for (Matches m : subMatches) {
+        MatchesIterator mi = m.getMatches(field);
+        if (mi != null) {
+          mis.add(mi);
+        }
+      }
+      matches.put(field, DisjunctionMatchesIterator.fromSubIterators(mis));
+    }
+    return new Matches(matches);
+  }
+
+  /**
+   * Create a {@link Matches} from a map of fields to iterators
+   */
+  protected Matches(Map<String, MatchesIterator> matches) {
+    this.matches = matches;
+  }
+
+  private Matches(String field, MatchesIterator iterator) {
+    this.matches = new HashMap<>();
+    this.matches.put(field, iterator);
+  }
+
+  /**
+   * Returns a {@link MatchesIterator} over the matches for a single field,
+   * or {@code null} if there are no matches in that field
+   */
+  public MatchesIterator getMatches(String field) {
+    return matches.get(field);
```

Comment:
I would expect this method to return a new iterator. But here if you call getMatches, then use it, and then call getMatches again, you will get an already consumed instance.